packages/std/src/math/fraction.rs

```diff
@@ -12,3 +12,129 @@ pub trait Fraction<T>: Sized {
+                self,
+                rhs: F,
+            ) -> Result<Self, CheckedMultiplyFractionError> {
+                let divisor = rhs.denominator().into();
+                let res = self
+                    .full_mul(rhs.numerator().into())
+                    .checked_div(divisor.into())?;
+                Ok(res.try_into()?)
+            }
+
+            pub fn mul_floor<F: Fraction<T>, T: Into<$Uint>>(self, rhs: F) -> Self {
+                self.checked_mul_floor(rhs).unwrap()
+            }
+
+            pub fn checked_mul_ceil<F: Fraction<T>, T: Into<$Uint>>(
+                self,
+                rhs: F,
+            ) -> Result<Self, CheckedMultiplyFractionError> {
+                let divisor = rhs.denominator().into();
+                let remainder = self
+                    .full_mul(rhs.numerator().into())
+                    .checked_rem(divisor.into())?;
+                let floor_result = self.checked_mul_floor(rhs)?;
+                if !remainder.is_zero() {
+                    Ok($Uint::one().checked_add(floor_result)?)
+                } else {
+                    Ok(floor_result)
+                }
+            }
+
+            pub fn mul_ceil<F: Fraction<T>, T: Into<$Uint>>(self, rhs: F) -> Self {
+                self.checked_mul_ceil(rhs).unwrap()
+            }
+
+            pub fn div_floor<F: Fraction<T>, T: Into<$Uint>>(self, rhs: F) -> Self
+            where
+                Self: Sized,
+            {
+                self.checked_div_floor(rhs).unwrap()
+            }
+
+            pub fn checked_div_floor<F: Fraction<T>, T: Into<$Uint>>(
+                self,
+                rhs: F,
+            ) -> Result<Self, CheckedMultiplyFractionError>
+            where
+                Self: Sized,
+            {
+                let divisor = rhs.numerator().into();
+                let res = self
+                    .full_mul(rhs.denominator().into())
+                    .checked_div(divisor.into())?;
+                Ok(res.try_into()?)
+            }
+
+            pub fn div_ceil<F: Fraction<T>, T: Into<$Uint>>(self, rhs: F) -> Self
+            where
+                Self: Sized,
+            {
+                self.checked_div_ceil(rhs).unwrap()
+            }
+
+            pub fn checked_div_ceil<F: Fraction<T>, T: Into<$Uint>>(
+                self,
+                rhs: F,
+            ) -> Result<Self, CheckedMultiplyFractionError>
+            where
+                Self: Sized,
+            {
+                let divisor = rhs.numerator().into();
+                let remainder = self
+                    .full_mul(rhs.denominator().into())
+                    .checked_rem(divisor.into())?;
+                let floor_result = self.checked_div_floor(rhs)?;
+                if !remainder.is_zero() {
+                    Ok($Uint::one().checked_add(floor_result)?)
+                } else {
+                    Ok(floor_result)
+                }
```

Comment:
This implementation is using https://stackoverflow.com/a/30824434/2013738, right? Now by using `checked_div_floor` here, the full mul as well as the widening into the larger types is duplicated. Can we inline this implementation with something like

```rust
// unsigned integers x, y of double the size of self
let x = self.full_mul(rhs.denominator().into());
let y = rhs.numerator().into();

// calculate x/y and x%y
```